File: backend/scripts/revalidate_prices.py
import argparse
import json
import os
import sys
from collections import defaultdict
from copy import deepcopy
from datetime import datetime, timezone
from pathlib import Path

from dotenv import load_dotenv

BACKEND_ROOT = Path(__file__).resolve().parents[1]
if str(BACKEND_ROOT) not in sys.path:
    sys.path.insert(0, str(BACKEND_ROOT))

from coin_cataloguer.price_validation import (
    assess_price_validation,
    fetch_bullion_rates,
    research_validated_price,
    search_market_context,
    should_research_item,
)


PROJECT_ROOT = Path(__file__).resolve().parents[2]
DATA_DIR = PROJECT_ROOT / "temp" / "data"
PROGRESS_PATH = PROJECT_ROOT / "temp" / "output" / "manual-price-revalidation-progress.json"
DEFAULT_COLLECTIONS = ["british", "mughals", "princely-states", "sultanate"]


def parse_args():
    parser = argparse.ArgumentParser(description="Manual price revalidation for catalog collections")
    parser.add_argument("--collections", nargs="*", default=DEFAULT_COLLECTIONS)
    parser.add_argument("--dry-run", action="store_true")
    parser.add_argument("--max-research", type=int, default=40)
    return parser.parse_args()


def now_iso():
    return datetime.now(timezone.utc).isoformat()


def write_progress(progress):
    PROGRESS_PATH.parent.mkdir(parents=True, exist_ok=True)
    PROGRESS_PATH.write_text(json.dumps(progress, indent=2, ensure_ascii=False))


def find_data_path(slug: str) -> Path:
    return DATA_DIR / f"{slug}.json"


def format_price_text(min_value: int, max_value: int) -> str:
    if min_value <= 0 and max_value <= 0:
        return ""
    if min_value == max_value:
        return f"{min_value:,}"
    return f"{min_value:,} - {max_value:,}"


def build_query(item):
    metadata = item.get("metadata", {})
    return " ".join(
        part
        for part in [
            item.get("title"),
            metadata.get("ruler_or_issuer") or metadata.get("rulerOrIssuer"),
            metadata.get("denomination"),
            metadata.get("mint_or_place") or metadata.get("mintOrPlace"),
            item.get("period"),
            "coin price",
        ]
        if part
    )


def apply_validation(item, assessment, rates, researched):
    metadata = item.setdefault("metadata", {})
    original_price = metadata.get("estimated_price_inr") or metadata.get("estimatedPriceInr") or ""
    if researched:
        min_value = int(researched.get("validated_price_min") or 0)
        max_value = int(researched.get("validated_price_max") or 0)
        if assessment["metal_value_inr"] > 0:
            min_value = max(min_value, assessment["metal_value_inr"])
            max_value = max(max_value, min_value)
        validated_text = researched.get("validated_price_text") or format_price_text(min_value, max_value)
        researched["validated_price_min"] = min_value
        researched["validated_price_max"] = max_value
        researched["validated_price_text"] = validated_text
    elif assessment["status"] == "below_metal_floor" and assessment["metal_value_inr"] > 0:
        floor_min = assessment["metal_value_inr"]
        floor_max = int(floor_min * 1.2)
        validated_text = format_price_text(floor_min, floor_max)
        researched = {
            "validated_price_min": floor_min,
            "validated_price_max": floor_max,
            "validated_price_text": validated_text,
            "confidence": "low",
            "manual_review_required": True,
            "reasoning": "Auto-adjusted to stay above the live bullion floor pending a fuller market review.",
            "search_results": [],
        }
    elif researched and researched.get("validated_price_text"):
        validated_text = researched["validated_price_text"]
    else:
        validated_text = original_price

    if original_price and original_price != validated_text and "original_estimated_price_inr" not in metadata:
        metadata["original_estimated_price_inr"] = original_price

    metadata["estimated_price_inr"] = validated_text
    metadata["price_validation_status"] = (
        "researched" if researched else assessment["status"]
    )
    metadata["price_validation_flags"] = assessment["flags"]
    metadata["price_validation_updated_at"] = now_iso()
    metadata["intrinsic_metal_value_inr"] = assessment["metal_value_inr"]
    metadata["bullion_reference"] = {
        "as_of": rates["as_of"],
        "gold_24k_inr_per_gram": rates["gold_24k_inr_per_gram"],
        "silver_999_inr_per_gram": rates["silver_999_inr_per_gram"],
    }
    if researched:
        metadata["price_validation_notes"] = researched.get("reasoning", "")
        metadata["price_validation_sources"] = [
            result["url"] for result in researched.get("search_results", [])
        ]
        metadata["price_validation_confidence"] = researched.get("confidence", "low")
        metadata["price_validation_manual_review_required"] = researched.get("manual_review_required", False)

    return item


def main():
    args = parse_args()
    load_dotenv(PROJECT_ROOT / "backend" / ".env")

    rates = fetch_bullion_rates()
    progress = {
        "started_at": now_iso(),
        "updated_at": now_iso(),
        "rates": rates,
        "collections": {},
        "totals": {
            "processed": 0,
            "updated": 0,
            "flagged": 0,
            "researched": 0,
        },
        "current": None,
    }
    write_progress(progress)

    researched_count = 0

    for slug in args.collections:
        data_path = find_data_path(slug)
        if not data_path.exists():
            continue

        payload = json.loads(data_path.read_text(encoding="utf-8"))
        items = payload.get("items", [])
        collection_stats = defaultdict(int)
        progress["collections"][slug] = {
            "total": len(items),
            "processed": 0,
            "updated": 0,
            "flagged": 0,
            "researched": 0,
        }

        updated_items = []
        for index, item in enumerate(items, start=1):
            metadata = item.get("metadata", {})
            assessment = assess_price_validation(
                materials=item.get("materials") or [],
                weight_text=metadata.get("weight_estimate") or metadata.get("weightEstimate"),
                estimated_price_text=metadata.get("estimated_price_inr") or metadata.get("estimatedPriceInr"),
                gold_24k_inr_per_gram=rates["gold_24k_inr_per_gram"],
                silver_999_inr_per_gram=rates["silver_999_inr_per_gram"],
            )

            researched = None
            if should_research_item(item=item, assessment=assessment) and researched_count < args.max_research:
                try:
                    query = build_query(item)
                    search_results = search_market_context(query)
                    researched = research_validated_price(
                        item=item,
                        rates=rates,
                        assessment=assessment,
                        search_results=search_results,
                    )
                    if researched:
                        researched["search_results"] = search_results
                        researched_count += 1
                        collection_stats["researched"] += 1
                except Exception as exc:
                    assessment["flags"] = [*assessment["flags"], "research_error"]
                    metadata = item.setdefault("metadata", {})
                    metadata["price_validation_notes"] = f"Research error: {exc}"

            updated_item = apply_validation(deepcopy(item), assessment, rates, researched)
            updated_items.append(updated_item)

            if assessment["flags"]:
                collection_stats["flagged"] += 1
            collection_stats["processed"] += 1
            collection_stats["updated"] += 1

            progress["collections"][slug].update(collection_stats)
            progress["totals"]["processed"] += 1
            progress["totals"]["updated"] += 1
            if assessment["flags"]:
                progress["totals"]["flagged"] += 1
            if researched:
                progress["totals"]["researched"] += 1

            progress["current"] = {
                "collection": slug,
                "index": index,
                "total": len(items),
                "item_id": item.get("id"),
                "title": item.get("title"),
            }
            progress["updated_at"] = now_iso()
            write_progress(progress)
            if index == 1 or index % 25 == 0 or index == len(items):
                print(
                    f"[{slug}] {index}/{len(items)} "
                    f"flagged={collection_stats['flagged']} researched={collection_stats['researched']}"
                )

        if not args.dry_run:
            payload["items"] = updated_items
            data_path.write_text(json.dumps(payload, indent=2, ensure_ascii=False), encoding="utf-8")

    progress["completed_at"] = now_iso()
    progress["current"] = None
    write_progress(progress)
    print(json.dumps(progress["totals"], indent=2))


if __name__ == "__main__":
    main()

File: backend/scripts/revalidatePrices.js
const fs = require('fs');
const path = require('path');
const admin = require('firebase-admin');

const PROJECT_ROOT = path.resolve(__dirname, '..', '..');
const DATA_DIR = path.join(PROJECT_ROOT, 'temp', 'data');
const OUTPUT_DIR = path.join(PROJECT_ROOT, 'temp', 'output');
const PROGRESS_PATH = path.join(OUTPUT_DIR, 'price-revalidation-progress.json');
const REPORT_PATH = path.join(OUTPUT_DIR, 'price-revalidation-report.json');
const SERVICE_ACCOUNT_PATHS = [
  path.join(PROJECT_ROOT, 'indian-heritage-gallery-firebase-adminsdk-fbsvc-1c3ae1c07a.json'),
  path.join(PROJECT_ROOT, 'temp', 'serviceAccountKey.json'),
];
const DATA_FILE_MAP = {
  british: 'british.json',
  mughals: 'mughals.json',
  'princely-states': 'princely-states.json',
  sultanate: 'sultanate.json',
};

function parseArgs() {
  const args = {
    collections: Object.keys(DATA_FILE_MAP),
    ratesFile: path.join(DATA_DIR, 'price-validation-rates-2026-04-03.json'),
    dryRun: false,
  };

  const argv = process.argv.slice(2);
  for (let index = 0; index < argv.length; index += 1) {
    if (argv[index] === '--collections' && argv[index + 1]) {
      args.collections = argv[index + 1].split(',').map((value) => value.trim()).filter(Boolean);
      index += 1;
    }
    if (argv[index] === '--rates-file' && argv[index + 1]) {
      args.ratesFile = path.resolve(argv[index + 1]);
      index += 1;
    }
    if (argv[index] === '--dry-run') {
      args.dryRun = true;
    }
  }

  return args;
}

function findServiceAccountPath() {
  return SERVICE_ACCOUNT_PATHS.find((candidate) => fs.existsSync(candidate)) || null;
}

function loadJson(filePath, fallback = null) {
  if (!fs.existsSync(filePath)) return fallback;
  return JSON.parse(fs.readFileSync(filePath, 'utf8'));
}

function writeJson(filePath, value) {
  fs.mkdirSync(path.dirname(filePath), { recursive: true });
  fs.writeFileSync(filePath, `${JSON.stringify(value, null, 2)}\n`, 'utf8');
}

function parseNumericValues(text) {
  return Array.from(
    String(text || '').matchAll(/(\d[\d,]*(?:\.\d+)?)\s*(k|thousand|lakh|lac|crore|cr|million|mn)?/gi),
  )
    .map((match) => {
      const numericValue = Number(match[1].replace(/,/g, ''));
      const suffix = String(match[2] || '').toLowerCase();
      if (!Number.isFinite(numericValue) || numericValue <= 0) return null;
      const multiplier =
        suffix === 'k' || suffix === 'thousand'
          ? 1000
          : suffix === 'lakh' || suffix === 'lac'
            ? 100000
            : suffix === 'crore' || suffix === 'cr'
              ? 10000000
              : suffix === 'million' || suffix === 'mn'
                ? 1000000
                : 1;
      return numericValue * multiplier;
    })
    .filter((value) => Number.isFinite(value) && value > 0);
}

function derivePriceRange(priceText) {
  const matches = parseNumericValues(priceText);
  if (matches.length === 0) {
    return { estimatedPriceMin: 0, estimatedPriceMax: 0, estimatedPriceAvg: 0 };
  }
  const estimatedPriceMin = Math.min(...matches);
  const estimatedPriceMax = Math.max(...matches);
  return {
    estimatedPriceMin,
    estimatedPriceMax,
    estimatedPriceAvg:
      matches.length === 1 ? matches[0] : Math.round((estimatedPriceMin + estimatedPriceMax) / 2),
  };
}

function deriveWeightGrams(weightText, fallbackWeightGrams = null) {
  const matches = parseNumericValues(weightText);
  if (matches.length > 0) return matches[0];
  return Number.isFinite(fallbackWeightGrams) && fallbackWeightGrams > 0 ? fallbackWeightGrams : null;
}

function roundToHundreds(value) {
  if (!Number.isFinite(value) || value <= 0) return 0;
  return Math.round(value / 100) * 100;
}

function formatCurrencyRange(min, max) {
  if (!Number.isFinite(min) || min <= 0) return '';
  const formatter = new Intl.NumberFormat('en-IN', { maximumFractionDigits: 0 });
  if (!Number.isFinite(max) || max <= 0 || Math.abs(max - min) < 1) {
    return `₹${formatter.format(roundToHundreds(min))}`;
  }
  return `₹${formatter.format(roundToHundreds(min))} - ₹${formatter.format(roundToHundreds(max))}`;
}

function extractPurity(text, defaultPurity) {
  const normalized = String(text || '').toLowerCase();
  const decimalMatch = normalized.match(/0\.(\d{3})/);
  if (decimalMatch) return Number(`0.${decimalMatch[1]}`);
  const finenessMatch = normalized.match(/\b(999|995|916|900|818|750|585|500)\b/);
  if (finenessMatch) return Number(finenessMatch[1]) / 1000;
  const percentMatch = normalized.match(/\b(\d{1,3})\s*%\b/);
  if (percentMatch) return Number(percentMatch[1]) / 100;
  if (normalized.includes('quaternary silver')) return 0.5;
  return defaultPurity;
}

function classifyPreciousMetal(item) {
  const materialText = [
    ...(Array.isArray(item.materials) ? item.materials : []),
    item.metadata?.material,
  ]
    .filter(Boolean)
    .join(' ')
    .toLowerCase();

  if (!materialText || /plated|debased|forgery|none detected|unable to determine|n\/a|none\b/.test(materialText)) {
    return null;
  }
  if (materialText.includes('gold')) {
    return { metal: 'gold', purity: extractPurity(materialText, 0.9) };
  }
  if (materialText.includes('silver') && !materialText.includes('billon')) {
    return { metal: 'silver', purity: extractPurity(materialText, 0.75) };
  }
  return null;
}

function buildValidation(item, rates) {
  const currentPriceText = String(item.metadata?.estimatedPriceInr || '').trim();
  const currentPriceRange = derivePriceRange(currentPriceText || item.estimatedPriceAvg);
  const weightGrams = deriveWeightGrams(item.metadata?.weightEstimate, item.weightGrams);
  const preciousMetal = classifyPreciousMetal(item);
  const flags = [];
  const notes = [];

  if (!preciousMetal || !weightGrams) {
    if (!currentPriceText && currentPriceRange.estimatedPriceAvg === 0) {
      flags.push('missing_price');
    }
    if (!weightGrams) {
      flags.push('missing_weight');
    }
    return {
      status: flags.length > 0 ? 'needs_manual_review' : 'validated_without_metal_floor',
      flags,
      notes,
      validatedPriceText: currentPriceText,
      validatedPriceMin: currentPriceRange.estimatedPriceMin,
      validatedPriceMax: currentPriceRange.estimatedPriceMax,
      validatedPriceAvg: currentPriceRange.estimatedPriceAvg,
      bullionFloorInr: null,
      preciousMetal,
      weightGrams,
      changed: false,
    };
  }

  const ratePerGram = preciousMetal.metal === 'gold' ? rates.gold999PerGramInr : rates.silver999PerGramInr;
  const bullionFloorInr = Math.round(weightGrams * ratePerGram * preciousMetal.purity);
  const conservativeMin =
    preciousMetal.metal === 'gold'
      ? roundToHundreds(bullionFloorInr * 1.1)
      : roundToHundreds(bullionFloorInr * 1.0);
  const conservativeMax =
    preciousMetal.metal === 'gold'
      ? roundToHundreds(bullionFloorInr * 1.35)
      : roundToHundreds(bullionFloorInr * 1.2);

  let validatedPriceMin = currentPriceRange.estimatedPriceMin;
  let validatedPriceMax = currentPriceRange.estimatedPriceMax;
  let validatedPriceAvg = currentPriceRange.estimatedPriceAvg;
  let validatedPriceText = currentPriceText;
  let status = 'validated_against_bullion_floor';
  let changed = false;

  if (!validatedPriceAvg) {
    flags.push('missing_price');
    notes.push('Filled missing price from conservative bullion-floor range.');
    validatedPriceMin = conservativeMin;
    validatedPriceMax = conservativeMax;
    validatedPriceAvg = Math.round((validatedPriceMin + validatedPriceMax) / 2);
    validatedPriceText = formatCurrencyRange(validatedPriceMin, validatedPriceMax);
    status = 'filled_from_bullion_floor';
    changed = true;
  } else if (validatedPriceMax < bullionFloorInr * 0.98) {
    flags.push('below_bullion_floor');
    notes.push('Existing estimate was below conservative bullion floor and has been repriced upward.');
    validatedPriceMin = conservativeMin;
    validatedPriceMax = Math.max(conservativeMax, roundToHundreds(validatedPriceMax * 1.5));
    validatedPriceAvg = Math.round((validatedPriceMin + validatedPriceMax) / 2);
    validatedPriceText = formatCurrencyRange(validatedPriceMin, validatedPriceMax);
    status = 'repriced_below_bullion_floor';
    changed = true;
  } else if (validatedPriceMin < bullionFloorInr * 0.98) {
    flags.push('min_price_below_bullion_floor');
    notes.push('Raised the lower bound to a conservative bullion-aware floor.');
    validatedPriceMin = conservativeMin;
    validatedPriceMax = Math.max(validatedPriceMax, conservativeMax);
    validatedPriceAvg = Math.round((validatedPriceMin + validatedPriceMax) / 2);
    validatedPriceText = formatCurrencyRange(validatedPriceMin, validatedPriceMax);
    status = 'raised_min_to_bullion_floor';
    changed = true;
  }

  return {
    status,
    flags,
    notes,
    validatedPriceText,
    validatedPriceMin,
    validatedPriceMax,
    validatedPriceAvg,
    bullionFloorInr,
    preciousMetal,
    weightGrams,
    changed,
  };
}

function createProgressPayload({ args, rates, summaries, startedAt, status, current }) {
  const totals = Object.values(summaries).reduce(
    (accumulator, summary) => {
      accumulator.totalItems += summary.totalItems;
      accumulator.processedItems += summary.processedItems;
      accumulator.updatedItems += summary.updatedItems;
      accumulator.flaggedItems += summary.flaggedItems;
      return accumulator;
    },
    { totalItems: 0, processedItems: 0, updatedItems: 0, flaggedItems: 0 },
  );

  return {
    status,
    startedAt,
    updatedAt: new Date().toISOString(),
    collections: args.collections,
    rates,
    totals,
    summaries,
    current,
  };
}

function syncLocalData(collectionSlug, updatesById) {
  const fileName = DATA_FILE_MAP[collectionSlug];
  if (!fileName) return;
  const filePath = path.join(DATA_DIR, fileName);
  const payload = loadJson(filePath);
  if (!payload || !Array.isArray(payload.items)) return;

  for (const item of payload.items) {
    const update = updatesById.get(item.id);
    if (!update) continue;
    item.metadata = item.metadata || {};
    item.metadata.estimated_price_inr = update.metadataEstimatedPriceInr;
    item.metadata.price_validation = update.localValidationPayload;
  }

  writeJson(filePath, payload);
}

async function main() {
  const args = parseArgs();
  const rates = loadJson(args.ratesFile);
  if (!rates) {
    throw new Error(`Rates file not found: ${args.ratesFile}`);
  }

  const serviceAccountPath = findServiceAccountPath();
  if (!serviceAccountPath) {
    throw new Error('Firebase service account key not found.');
  }

  admin.initializeApp({
    credential: admin.credential.cert(require(serviceAccountPath)),
  });
  const db = admin.firestore();
  const startedAt = new Date().toISOString();
  const summaries = {};
  const report = {
    rates,
    startedAt,
    finishedAt: null,
    collections: {},
    flaggedItems: [],
  };

  for (const collectionSlug of args.collections) {
    console.log(`Revalidating ${collectionSlug}...`);
    const snapshot = await db.collection('items').where('collectionSlug', '==', collectionSlug).get();
    const items = snapshot.docs.map((docSnapshot) => ({ id: docSnapshot.id, ...docSnapshot.data() }));
    const updatesById = new Map();
    const summary = {
      totalItems: items.length,
      processedItems: 0,
      updatedItems: 0,
      flaggedItems: 0,
      status: 'running',
    };
    summaries[collectionSlug] = summary;

    let batch = db.batch();
    let batchCount = 0;

    for (const item of items) {
      const validation = buildValidation(item, rates);
      const current = { collectionSlug, itemId: item.id, title: item.title };
      summary.processedItems += 1;
      if (validation.flags.length > 0) {
        summary.flaggedItems += 1;
        report.flaggedItems.push({
          collectionSlug,
          itemId: item.id,
          title: item.title,
          status: validation.status,
          flags: validation.flags,
          currentPrice: item.metadata?.estimatedPriceInr || '',
          validatedPrice: validation.validatedPriceText,
          bullionFloorInr: validation.bullionFloorInr,
        });
      }

      const priceValidation = {
        status: validation.status,
        flags: validation.flags,
        notes: validation.notes,
        bullionFloorInr: validation.bullionFloorInr,
        metalType: validation.preciousMetal?.metal || null,
        metalPurity: validation.preciousMetal?.purity || null,
        weightGrams: validation.weightGrams,
        validatedPriceMin: validation.validatedPriceMin,
        validatedPriceMax: validation.validatedPriceMax,
        validatedPriceAvg: validation.validatedPriceAvg,
        validatedPriceText: validation.validatedPriceText,
        ratesAsOf: rates.asOf,
        ratesSource: rates.sources,
        methodology: 'deterministic-bullion-floor-v1',
        updatedAt: new Date().toISOString(),
      };

      const updatePayload = {
        estimatedPriceMin: validation.validatedPriceMin,
        estimatedPriceMax: validation.validatedPriceMax,
        estimatedPriceAvg: validation.validatedPriceAvg,
        metadata: {
          ...(item.metadata || {}),
          estimatedPriceInr: validation.validatedPriceText || item.metadata?.estimatedPriceInr || '',
        },
        priceValidation,
        updatedAt: admin.firestore.FieldValue.serverTimestamp(),
      };

      updatesById.set(item.id, {
        metadataEstimatedPriceInr: updatePayload.metadata.estimatedPriceInr,
        localValidationPayload: {
          status: validation.status,
          flags: validation.flags,
          bullion_floor_inr: validation.bullionFloorInr,
          validated_price_min: validation.validatedPriceMin,
          validated_price_max: validation.validatedPriceMax,
          validated_price_avg: validation.validatedPriceAvg,
          validated_price_text: validation.validatedPriceText,
          rates_as_of: rates.asOf,
        },
      });

      if (validation.changed) {
        summary.updatedItems += 1;
      }

      if (!args.dryRun) {
        const docRef = db.collection('items').doc(item.id);
        batch.set(docRef, updatePayload, { merge: true });
        batchCount += 1;
      }

      if (batchCount >= 300) {
        await batch.commit();
        batch = db.batch();
        batchCount = 0;
      }

      if (summary.processedItems % 25 === 0 || summary.processedItems === summary.totalItems) {
        writeJson(
          PROGRESS_PATH,
          createProgressPayload({
            args,
            rates,
            summaries,
            startedAt,
            status: 'running',
            current,
          }),
        );
        console.log(
          `  ${collectionSlug}: ${summary.processedItems}/${summary.totalItems} processed, ${summary.updatedItems} repriced, ${summary.flaggedItems} flagged`,
        );
      }
    }

    if (!args.dryRun && batchCount > 0) {
      await batch.commit();
    }

    if (!args.dryRun) {
      syncLocalData(collectionSlug, updatesById);
    }

    summary.status = 'completed';
    report.collections[collectionSlug] = summary;
  }

  report.finishedAt = new Date().toISOString();
  writeJson(REPORT_PATH, report);
  writeJson(
    PROGRESS_PATH,
    createProgressPayload({
      args,
      rates,
      summaries,
      startedAt,
      status: 'completed',
      current: null,
    }),
  );

  console.log(`Progress written to ${PROGRESS_PATH}`);
  console.log(`Report written to ${REPORT_PATH}`);
}

main().catch((error) => {
  writeJson(
    PROGRESS_PATH,
    {
      status: 'failed',
      updatedAt: new Date().toISOString(),
      error: error.message || String(error),
    },
  );
  console.error(error.message || error);
  process.exit(1);
});
